lib/kubeclient_next/client/api_builder.rb

```diff
@@ -39,18 +49,88 @@ def build!
-        client.define_singleton_method("get_#{resource_description.plural_name}".to_sym) do |kwargs = {}|
-          namespace = kwargs[:namespace] if resource_description.namespaced
-          rest_client.get(resource_description.path_for_resources(namespace: namespace))
+      def define_create_resource(client, rest_client, resource_description)
+        client.define_singleton_method("create_#{resource_description.singular_name}".to_sym) do |kwargs = {}|
+          namespace = kwargs.fetch(:namespace) if resource_description.namespaced
+          data = kwargs.fetch(:data)
+          headers = kwargs.fetch(:headers, {})
+          rest_client.post(resource_description.path_for_resources(namespace: namespace),
+            data: JSON.dump(data), headers: headers)
+        end
+      end
+
+      def define_delete_resource(client, rest_client, resource_description)
+        client.define_singleton_method("delete_#{resource_description.singular_name}".to_sym) do |kwargs = {}|
+          namespace = kwargs.fetch(:namespace) if resource_description.namespaced
+          name = kwargs.fetch(:name)
+          headers = kwargs.fetch(:headers, {})
+          rest_client.delete(resource_description.path_for_resource(namespace: namespace, name: name),
+            headers: headers)
         end
       end
 
       def define_get_resource(client, rest_client, resource_description)
         client.define_singleton_method("get_#{resource_description.singular_name}".to_sym) do |kwargs = {}|
-          namespace = kwargs[:namespace] if resource_description.namespaced
+          namespace = kwargs.fetch(:namespace) if resource_description.namespaced
           name = kwargs.fetch(:name)
-          rest_client.get(resource_description.path_for_resource(namespace: namespace, name: name))
+          headers = kwargs.fetch(:headers, {})
+          rest_client.get(resource_description.path_for_resource(namespace: namespace, name: name), headers: headers)
+        end
+      end
+
+      def define_get_resources(client, rest_client, resource_description)
+        client.define_singleton_method("get_#{resource_description.plural_name}".to_sym) do |kwargs = {}|
+          namespace = kwargs.fetch(:namespace) if resource_description.namespaced
+          headers = kwargs.fetch(:headers, {})
+          rest_client.get(resource_description.path_for_resources(namespace: namespace), headers: headers)
+        end
+      end
+
+      def define_patch_and_apply_resource(client, rest_client, resource_description)
+        define_patch_resource(client, rest_client, resource_description)
+        define_apply_resource(client, rest_client, resource_description)
+      end
+
+      def define_patch_resource(client, rest_client, resource_description)
+        # Lexically scope this method in the block to make it available inside the closure when invoked by client
+        scoped_content_type_for_patch_strategy = proc { |strategy| content_type_for_patch_strategy(strategy) }
```

Comment:
While this is certainly a little bit ugly, I chose to do it this way for 2 reasons:

- I didn't want to have the invocation need to explicitly reference the full namespace path, which needs to be done since we can't guarantee that the `client` param will always be an instance `Client::Client`, and it seems clumsy. For a similar reason, having it as a class method on APIBuilder seemed equally clumsy. Said that, I don't have a strong feeling either way.